DOC: Update numpy4matlab

cooperrc opened the pull request:
<!-- Please be sure you are following the instructions in the dev guidelines
http://www.numpy.org/devdocs/dev/development_workflow.html
-->
This PR closes #17094 that requests an update to the "NumPy for MATLAB users" restructured text document. 
Comments were integrated from @melissawm, @bjnath, @rossbar, @8bitmp3, and @eric-wieser 

Here is a brief list of changes:

1. Some minor consistencies and grammar were corrected (e.g. Matlab -> MATLAB throughout)
2. The discussion of `array` vs `matrix` was moved to the end of the document, since the general recommendation is to use `array`
3. New links were added in the benefits of NumPy including other SciPy/Matplotlib/etc. projects and a link to a wikipedia article with scientific software that uses Python as a scripting language
4. Some examples in the table were updated and some added e.g. `for` and `if` statements were added and `random` has the updated `default_rng` recommendation. 

<!-- We'd appreciate it if your commit message is properly formatted
http://www.numpy.org/devdocs/dev/development_workflow.html#writing-the-commit-message
-->

cooperrc committed:
first draft of numpy-for-matlab-users updates

cooperrc committed:
updated advantage for Python with scripting SDKs

cooperrc committed:
minor edits for review

cooperrc committed:
minor edits for review

cooperrc committed:
respond to comments @eric-weiser L28-36 and L46-49

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -43,142 +44,27 @@ Some Key Differences
 
    * - MATLAB®'s scripting language was created for doing linear algebra.
-       The syntax for basic matrix operations is nice and clean, but the API
-       for adding GUIs and making full-fledged applications is more or less
+       MATLAB's syntax for some array manipulations is more compact than
+       NumPy's.The syntax for basic matrix operations is nice and clean, but the
+       API for adding GUIs and making full-fledged applications is more or less
```
Missing space after the full stop. Also, the first sentence and first half of the second sentence would be better combined.

cooperrc replied:
Thanks for catching that. I could go either way with combining the sentences. My preference is shorter sentences, but I'll combine and let's take another look.

eric-wieser replied:
Looks good now, thanks.

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,11 +25,12 @@ Some Key Differences
-       arrays and return such arrays.  Operations on the 2-D instances of
-       these arrays are designed to act more or less like matrix operations
-       in linear algebra.
+   * - In MATLAB®, you can create multidimensional arrays. By default,
+       MATLAB® creates two dimensional arrays of double precision floating point
+       numbers.  Most expressions take such arrays and return arrays.
+       Operations on the 2-D instances of these arrays are designed to act 
+       like matrix operations in linear algebra.
+
      - In NumPy the basic type is a multidimensional ``array``.  Operations
        on these arrays in all dimensionalities including 2D are element-wise
        operations.  One needs to use specific functions for linear algebra
```
The way this was worded before was nice, because the sentence had similar structure in both columns. Can you try to bring the two cells into a similar sentence structure?

cooperrc replied:
That's a good idea, I'll make make the structure similar in both. The main edit here was to highlight

* Matlab default = 2D arrays
* NumPy default = 0D or 1D arrays

eric-wieser replied:
I'd say perhaps more relevant is:

* Matlab: 2 or more dimensional arrays, default double
* Numpy: n-dimensional arrays, default varies

Lets see what others think.

cooperrc replied:
> I'd say perhaps more relevant is:
> 
> * Matlab: 2 or more dimensional arrays, default double
> * Numpy: n-dimensional arrays, default varies

I like that. In this case, one of the benefits of NumPy over MATLAB is the control over the shapes and types of your arrays. So maybe it would be good to call out that the `array` function defaults to your data types and dimensions. 

I'm thinking about 2 inputs of MATLAB vs Python here:

1. A single integer input:
```matlab
> a=1
> shape(a)
ans =
  1 1
> class(a)
  ans = double
```
vs 
```python
>>> a=np.array(1)
>>> a.shape
  ()
>>> a.dtype
  dtype('int64')
```
2. A list of integers:
```matlab
> a=[1,2,3]
> shape(a)
ans = 
  1 3
> class(a)
  ans = double
```
vs
```python
>>> a=np.array([1,2,3])
>>> a.shape
(3,)
>>> a.dtype
  dtype('int64')
```

eric-wieser replied:
That definitely looks like something which would be useful to have somewhere on the page.

Perhap the matlab "list of integers" example should use `int8([1, 2, 3])` or whatever the matlab syntax is, for fairness.

cooperrc replied:
Maybe this example should go in the table below. The equivalent for matlab signed integers would be `int64([1,2,3])`.

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -627,47 +578,167 @@ Dijkstra <https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html
 reshape(x,3,4); should become z = x.reshape(3,4,order='F').copy() in
 NumPy.
 
+'array' or 'matrix'? Which should I use?
```
Was this section edited or just moved?

cooperrc replied:
This was moved, not edited.

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -206,21 +92,38 @@ General Purpose Equivalents
+
+     - ::
+
+         for i in range(1,4):
```
Let's try to encourage PEP8 while we're translating,
```suggestion
         for i in range(1, 4):
```

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -231,13 +134,39 @@ General Purpose Equivalents
+
+     - ::
+
+         a=1
+         if a==0:
+             print('a=0')
+         elif a==1:
+             print('a=1')
```
PEP8
```suggestion
         a = 1
         if a == 0:
             print('a=0')
         elif a == 1:
             print('a=1')
```

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -429,16 +358,28 @@ Linear Algebra Equivalents
-     - square diagonal matrix whose nonzero values are the elements of
-       ``a``
+         from numpy.random import default_rng
+         rng=default_rng(42)
```
PEP8
```suggestion
         rng = default_rng(42)
```

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,31 +479,38 @@ Linear Algebra Equivalents
+     - return the rank of 2D array, ``a``
+
+   * - ``c=chol(a)`` where ``a==c'*c``
+     - ``c=linalg.cholesky(a)`` where ``a==c@c.T``
```
PEP8
```suggestion
     - ``c = linalg.cholesky(a)`` where ``a == c@c.T``
```

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -576,28 +525,30 @@ Linear Algebra Equivalents
+     - ``sort(a)`` or ``a.sort(axis=0)``
+     - sort each column of a 2D matrix, ``a``
 
-   * - ``[b,I] = sortrows(a,i)``
```
I think this one is still worth keeping

cooperrc replied:
I can put this back in, but the description needs an update. This is not sorting rows, its sorting the entire array by row based upon a column number, `i`.  e.g.

```matlab
A=[1,2,3;3,2,1;2,3,1]
sortrows(A,1)
ans =

   1   2   3
   2   3   1
   3   2   1
```
vs

```python
>>> A = np.array([[1,2,3],[3,2,1],[2,3,1]])
>>> I = argsort(A[:,0])
>>> A[I,:]
    array([[1, 2, 3],
           [2, 3, 1],
           [3, 2, 1]])
```

cooperrc replied:
I added the `sortrows` for matlab again, but distinguished it from a typical sort:

* ``[b,I]=sortrows(a,1)``
* ``I=argsort(a[:,0]); b=a[I,:]``

`save the array ``a`` as array ``b`` with rows sorted by the first column`

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -356,12 +285,12 @@ Linear Algebra Equivalents
 
    * - ``find(a>0.5)``
-     - ``nonzero(a>0.5)``
+     - ``where(a>0.5)``
```
```suggestion
     - ``nonzero(a>0.5)``
```
The overloading of `where` just confuses users. Furthermore, the docs for `where` advise preferring `nonzero` anyway.

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,160 +25,46 @@ Some Key Differences
+       matrix operations in linear algebra. Most expressions take such arrays and
+       return arrays.
+     - In NumPy, the basic type is a multidimensional ``array``. By default,
+       NumPy creates 0-D or 1-D arrays for inputs of numbers or lists,
```
Perhaps confusing, as `np.array([[1]])` is not 1D, even though it is a list.

cooperrc replied:
`np.array([[1]]).shape` is `(1,1)` so it is a 2D array. If you assign it to the variable `a=np.array([[1]])`, you can reference the `1` with `a[0,0]`.

This is a tripping point for Matlab users migrating to NumPy. Because in Matlab you would get the number 1 by calling 

`a`, `a(1)`, or `a(1,1)`, 

but in NumPy if you're trying to call the number `1`, you need to call out the value in the multidimensional array

`a[0,0]` -> the number `1`

`a[0]` -> a 1D array with one value, `1`

a -> a 2D array with one value, `1`

Maybe we need a little more details in this description.

cooperrc replied:
@eric-wieser, I've got 88 lines of rst (see below) that I could add to walk a matlab user through this idea of multidimensional arrays in numpy. I think it would fit in the Introduction at line 16. I pasted the code in a gist to see what others think. 

[MATLAB vs NumPy array creation defaults](https://gist.github.com/cooperrc/e023ee0cdf125b482a131fa992f329a7)

eric-wieser replied:
As a quick fix for this line then, perhaps:
```suggestion
       NumPy creates 0-D or 1-D arrays for inputs of numbers or lists of numbers,
```

cooperrc replied:
looks good to me, I'll make the change

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,160 +25,46 @@ Some Key Differences
+       is access to Python libraries including: `SciPy
+       <https://www.scipy.org/>`_, `Matplotlib <https://matplotlib.org/>`_,
+       `Pandas <https://pandas.pydata.org/>`_, `OpenCV <https://opencv.org/>`_,
+       and more. In addition to Python-specific packages, many `scientific
+       computing softwares
```
```suggestion
       and more. In addition to Python-specific packages, much `scientific
       computing software
```

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

PEP8

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

PEP8

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

PEP8

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

PEP8

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

`where` points to `nonzero` so revert to `nonzero`

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -231,13 +134,39 @@ General Purpose Equivalents
+   * - ``load data.mat``
+     - ``scipy.io.loadmat('data.mat')``
+     - load variables saved to the file ``data.mat`` note, if saving files from
+       MATLAB/Octave use a recent binary format. Python's ``loadmat`` will
```
Punctuation seems off here

eric-wieser replied:
```suggestion
       MATLAB/Octave use a recent binary format, Python's ``loadmat`` will
```

cooperrc replied:
updated the punctuation in discussion column

cooperrc committed:
add sortrows discussion and distinguish from sort

cooperrc committed:
L165- fix load *.mat punctuation

cooperrc committed:
L165- fix load *.mat punctuation

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

plural -> singular softwares -> software

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,31 +479,38 @@ Linear Algebra Equivalents
-     - ``linalg.cholesky(a).T``
-     - cholesky factorization of a matrix (``chol(a)`` in matlab returns an
+   * - ``rank(a)``
+     - ``matrix_rank(a)``
```
```suggestion
     - ``linalg.matrix_rank(a)``
```

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

punctuation correctio

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

specify matrix_rank(a) as linalg.matrix_rank(a)
not imported from scipy.linalg

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,31 +479,38 @@ Linear Algebra Equivalents
 
-   * - ``chol(a)``
-     - ``linalg.cholesky(a).T``
-     - cholesky factorization of a matrix (``chol(a)`` in matlab returns an
```
Cholesky should be capitalized.

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,31 +479,38 @@ Linear Algebra Equivalents
      - ``U, S, Vh = linalg.svd(a), V = Vh.T``
      - singular value decomposition of ``a``
 
```
`rank(a)` <-> `linalg.matrix_rank(a)` appears twice in the table

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,31 +479,38 @@ Linear Algebra Equivalents
-     - LU decomposition (note: P(Matlab) == transpose(P(numpy)) )
+   * - ``[L,U,P]=lu(a)`` where ``a==P'*L*U``
+     - ``P,L,U = scipy.linalg.lu(a)`` where ``a==P@L@U``
+     - LU decomposition (note: P(MATLAB) == transpose(P(NumPy)) )
```
Extra space before the last parenthesis.

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -627,47 +582,167 @@ Dijkstra <https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html
+code which relies on the scan order, then this MATLAB code: z =
 reshape(x,3,4); should become z = x.reshape(3,4,order='F').copy() in
 NumPy.
 
```
Here and above, some of the code (e.g. "z = reshape(x, 3, 4)") is not surrounded by backticks, which makes the rendering less clear for those reading. Would it be possible to review the document and add backticks to inline code expressions so they are rendered in code-mode?

cooperrc replied:
I went through these paragraphs and found the inline code. Glad you caught those.

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -627,47 +582,167 @@ Dijkstra <https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html
+**Use arrays**.
+
+-  They support multidimensional array algebra that is supported in MATLAB
+-  They are the standard vector/matrix/tensor type of NumpY. Many NumPy
```
NumpY -> NumPy

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -718,4 +801,9 @@ cross-reference.
 An extensive list of tools for scientific work with python can be
```
Python should be capitalized.

melissawm reviewed (changes requested):
Thanks for that @cooperrc ! I left a number of comments, some of them very minor, some only suggestions. One think I'd like to point out is that there are several pieces of Python code in that document non-PEP8 compliant, mainly around slicing and indexing operations. 

There are also some places where MATLAB has a registered symbol and other places where it doesn't. Might be worthwile mentioning in the beginning but dropping it for the rest of the text? I'm not sure how that can be done correctly. 

I also liked your extra document about creating arrays, and I think it's definitely worth adding. There are also some details to be fixed there (like PEP8 spacing and some wording).

Thanks again for doing this, I know it's a large document with many details but I think this is a very important document for those looking to make the switch.

bjnath commented:
> places where MATLAB has a registered symbol and other places where it doesn't

@melissawm Thanks for pointing this out.

[Google style guide](https://developers.google.com/style/trademarks) says

> For trademark marking or attribution in documentation, follow any usage guidelines provided by the owners of the respective marks.

[MathWorks](https://www.mathworks.com/content/dam/mathworks/mathworks-dot-com/support/books/members/pdfs/Trademark_Logo_Guidelines.pdf) says

>  use the relevant trademark symbol to mark the first occurrences

and 

> The word MATLAB®should appear in all capital letters.

bjnath commented:
> I pasted the code in a gist

If you intend it as part of this PR, please go ahead and include it.

melissawm commented:
> > I pasted the code in a gist
> 
> If you intend it as part of this PR, please go ahead and include it.

One option we can consider is splitting this document into two: one more of a "cheatsheet", mostly containing the table currently in this document, and another with more details and examples of how to do things properly in NumPy. This can be considered if the document get too large or if we feel there may be two separate audiences (I suspect that might be the case). But this can be a next step after this PR is merged.

eric-wieser commented:
> One option we can consider is splitting this document into two: one more of a "cheatsheet", mostly containing the table currently in this document, and another with more details and examples of how to do things properly in NumPy

I like this idea, especially if cells in the former link to the latter.

Although a single document would work well too now that we have nice sidebar navigation, as long as the table comes first.

cooperrc commented:
> > I pasted the code in a gist
> 
> If you intend it as part of this PR, please go ahead and include it.

I wasn't sure if it directly closes the issue of updating the current document. It would certainly help a Matlab user, but the original issue was just to update the original document. This would add new material. 

What do you think?

cooperrc commented:
> Thanks for that @cooperrc ! I left a number of comments, some of them very minor, some only suggestions. One think I'd like to point out is that there are several pieces of Python code in that document non-PEP8 compliant, mainly around slicing and indexing operations. 

Thanks for reviewing and the careful edits. It is very much appreciated!

> There are also some places where MATLAB has a registered symbol and other places where it doesn't. Might be worthwile mentioning in the beginning but dropping it for the rest of the text? I'm not sure how that can be done correctly. 

@bjnath found the correct way. Thanks!

> 
> I also liked your extra document about creating arrays, and I think it's definitely worth adding. There are also some details to be fixed there (like PEP8 spacing and some wording).

I have missed a lot of PEP8 formatting. Thanks for catching those. 

> Thanks again for doing this, I know it's a large document with many details but I think this is a very important document for those looking to make the switch.

You're very welcome, I agree. This is a document I have referenced many times converting between NumPy and MATLAB. I'm happy to give back.

cooperrc commented:
> > > I pasted the code in a gist
> > 
> > If you intend it as part of this PR, please go ahead and include it.
> 
> One option we can consider is splitting this document into two: one more of a "cheatsheet", mostly containing the table currently in this document, and another with more details and examples of how to do things properly in NumPy. This can be considered if the document get too large or if we feel there may be two separate audiences (I suspect that might be the case). But this can be a next step after this PR is merged.

@melissawm and @eric-wieser I like this idea too. It may even be worthwhile to consider a short tutorial on converting MATLAB to NumPy. We could pick an example that highlights some of the strengths in using NumPy over MATLAB. 

One thing that comes to mind is multidimensional arrays. In NumPy, you can nest lists and build N-D arrays (great for continuum mechanics and other tensor-based models), but in Matlab you have to recursively redefine the array to get more dimensions e.g.

```matlab
A=[1,2;3,4]
A(:,:,2)=[5,6;7,8]
```

Vs NumPy

```python
A=np.array([[[1,2],[3,4]],
    [[5,6],[7,8]]])
```

bjnath commented:
@cooperrc It might be worth making a commit now that incorporates @melissawm 's changes so we have a clean base for further review while we think about 2 vs 1.

cooperrc committed:
respond to @melissawm comments

cooperrc committed:
respond to @melissawm comments L63-remove actually

cooperrc commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -206,21 +92,38 @@ General Purpose Equivalents
 
    * - **MATLAB**
-     - **numpy**
+     - **NumPy**
```
That's true, we shouldn't encourage `import *`. 

I think it might be better to add `np.<func>` to the table. It will also help to separate `scipy.linalg` from `numpy.linalg` since there are overlapping functions.

cooperrc replied:
Latest commit removed the `import *` on L82. 

replaced those lines with 

```python
    import numpy as np
    from scipy import io, integrate, linalg, signal
    from scipy.sparse.linalg import eigs
```
commit 2812833bc5f1bb214ae86045e4582ceeb4b464dd

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,51 @@ Some Key Differences
-       things that Matlab just cannot, for instance dealing properly with
-       stacks of matrices.
-
-   * - In MATLAB®, arrays have pass-by-value semantics, with a lazy
```
The removal of `®` from these lines makes this a little harder to review. Would you be willing to do a very quick first-pass PR that does nothing but clean up the `®` symbols?

cooperrc replied:
Sure thing, I'll make another branch, "matlabcopyright", and submit the PR.

cooperrc replied:
Update: submitted PR #17215 that only removes the excess (R) symbols from MATLAB mentions. The only one left is the first mention as per @bjnath's reference.

cooperrc committed:
respond to @melissawm comments L82- remove import * and import functions

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,51 @@ Some Key Differences
-       in python 3.5 and above).
-
-   * - MATLAB® uses 1 (one) based indexing. The initial element of a
+   * - In MATLAB, the basic type is a multidimensional array. By default, MATLAB
```
```suggestion
   * - In MATLAB, the basic nonscalar type is a multidimensional array. By default, MATLAB
```

cooperrc replied:
Matlab defaults to a 2D array even for scalar assignments. So when you say `a=1`, it creates a 1x1 array and assigns it to `a`. You can reference the `1` by using `a(1)` or `a(1,1)`. 

I don't think matlab has a "scalar" assignment option because all variables are given dimensions. It does consider a 1x1 matrix a scalar, so

```matlab
>> a=1
>> isscalar(a)
TF =
    1
>> isscalar(a(1,1))
TF =
    1
```

This is kind of a strange convention in matlab.

bjnath replied:
Interesting! 
```
- In MATLAB, the basic type, even for scalars, is a multidimensional array.
```
Ordinarily I wouldn't tell MATLAB users what they already know, but this behavior is going to be contrasted with NumPy's.

cooperrc replied:
I like that change, @bjnath . 

I'll add the `even for scalars`. MATLAB's choice of everything is a matrix makes some things easier, but if you're trying to convert your code to NumPy, there are a few hiccups like this that come up. 

I don't think most MATLAB users think about the difference between scalar/vector/matrix in normal operations because it all works out seemlessly.

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,51 @@ Some Key Differences
-
-   * - MATLAB® uses 1 (one) based indexing. The initial element of a
+   * - In MATLAB, the basic type is a multidimensional array. By default, MATLAB
+       creates two dimensional arrays of double precision floating point numbers.
```
I'm not a hyphen fanatic, but I think most readers would expect to see one here.
```suggestion
       creates two-dimensional arrays of double precision floating point numbers.
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,51 @@ Some Key Differences
-   * - MATLAB® uses 1 (one) based indexing. The initial element of a
+   * - In MATLAB, the basic type is a multidimensional array. By default, MATLAB
+       creates two dimensional arrays of double precision floating point numbers.
+       Operations on the 2D instances of these arrays are designed to act like
+       matrix operations in linear algebra. Most expressions take such arrays and
+       return arrays.
```
```suggestion
       Operations on the 2D instances of these arrays are designed to act like
       matrix operations in linear algebra. Most expressions take such arrays and
       return arrays.
```
The second sentence doesn't add anything.
```suggestion
       Operations on the 2D instances of these arrays are modeled on
       matrix operations in linear algebra. 
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,51 @@ Some Key Differences
+       matrix operations in linear algebra. Most expressions take such arrays and
+       return arrays.
+     - In NumPy, the basic type is a multidimensional ``array``. By default,
+       NumPy creates 0-D or 1-D arrays for inputs of numbers or lists,
```
I don't understand "by default." Creating a NumPy array requires calling a routine, which means you have to make some choice.

seberg replied:
From the sideline: It probably wants to say, unlike matlab where matrices always have two dimensions, NumPy will create 0-D or 1-D arrays unless `ndmin=2` is passed.

cooperrc replied:
Good point, I was trying to keep the sentence structure the same the table:

"By default, MATLAB creates two-dimensional arrays of double precision floating point numbers."

and

"By default, NumPy creates 0-D or 1-D arrays for inputs of numbers or lists, respectively."

How about this instead:

"Array assignments in MATLAB are stored as two-dimensional arrays of double precision floating point numbers, unless you specify the number of dimensions and type."

and

"Array assignments in NumPy are stored as N-dimensional arrays with the minimum type required to hold the objects in the sequence, unless you specify the number of dimensions and type."

bjnath replied:
> the minimum type required
Not always true. linspace, for instance, uses floats even when it can use ints. Maybe just "usually" will fix it.

cooperrc replied:
> This seems like a deft fix, but
> 
> > the minimum type required
> 
> is not always true. linspace, for instance, uses floats even when it can use ints. Maybe just "usually" will fix it.
> 
> Also
> 
> > two-dimensional
> 
> For consistency at least in this doc, I think we should go with the numeral 2.

OK, change was pushed

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,51 @@ Some Key Differences
+       matrix operations in linear algebra. Most expressions take such arrays and
+       return arrays.
+     - In NumPy, the basic type is a multidimensional ``array``. By default,
+       NumPy creates 0-D or 1-D arrays for inputs of numbers or lists,
```
```suggestion
       NumPy creates 0-D or 1-D arrays for inputs of numbers or lists,
```
Here we have hyphens for dimensions; elsewhere we have `2D` with none.  I think I vote for no hyphens, but either way we should be consistent.

cooperrc replied:
hyphens were removed, thanks for catching that

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

add hyphen to L29

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

remove redundant sentence

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,51 @@ Some Key Differences
+       return arrays.
+     - In NumPy, the basic type is a multidimensional ``array``. By default,
+       NumPy creates 0-D or 1-D arrays for inputs of numbers or lists,
+       respectively.  Operations on these arrays in all dimensionalities
+       including 2D are element-wise operations.  One needs to use specific
+       functions for linear algebra (though for matrix multiplication, one can
+       use the ``@`` operator in Python 3.5 and above).
```
```suggestion
       respectively.  NumPy performs operations element-by-element, so multiplying
       2D arrays with `*` is not a matrix multiplication -- it's an element-by-element 
       multiplication. (The `@` operator, available since Python 3.5, can be used
       for conventional matrix multiplication.)
       functions for linear algebra (though for matrix multiplication, one can
       use the ``@`` operator in Python 3.5 and above).
```

eric-wieser replied:
Double backticks for code (operators here) in rst @bjnath

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,50 @@ Some Key Differences
+   * - MATLAB uses 1 (one) based indexing. The initial element of a
        sequence is found using a(1).
        :ref:`See note INDEXING <numpy-for-matlab-users.notes>`
      - Python uses 0 (zero) based indexing. The initial element of a
        sequence is found using a[0].
```
- NumPy, like Python, numbers indexes from 0; a[0] is the first element.

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -9,9 +9,9 @@ Introduction
 
 MATLAB® and NumPy/SciPy have a lot in common. But there are many
 differences. NumPy and SciPy were created to do numerical and scientific
-computing in the most natural way with Python, not to be MATLAB® clones.
+computing in the most natural way with Python, not to be MATLAB clones.
 This page is intended to be a place to collect wisdom about the
-differences, mostly for the purpose of helping proficient MATLAB® users
+differences, mostly for the purpose of helping proficient MATLAB users
 become proficient NumPy and SciPy users.
```
Suggested wording:

MATLAB® and NumPy have a lot in common, but NumPy  was created to
work with Python, not to be a MATLAB clone. This guide will help MATLAB 
users get started with NumPy.

(Unable to submit this as a "suggested change"; GitHub says "Applying suggestions on deleted lines is not supported.")

cooperrc replied:
I incorporated the change.

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,50 @@ Some Key Differences
-       an afterthought.
-     - NumPy is  based on Python, which was designed from the outset to be
-       an excellent general-purpose programming language.  While Matlab's
+   * - MATLAB's scripting language was created for doing linear algebra so the
```
```suggestion
   * - MATLAB's scripting language was created for linear algebra so the
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,50 @@ Some Key Differences
-
-   * - In MATLAB®, arrays have pass-by-value semantics, with a lazy
-       copy-on-write scheme to prevent actually creating copies until they
+       NumPy's. The syntax for basic matrix operations is nice and clean, but the
+       API for adding GUIs and making full-fledged applications is more or less
+       an afterthought.
```
```suggestion
       NumPy's. On the other hand, the API for adding GUIs and creating 
       full-fledged applications is more or less an afterthought.
```

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

change "doing linear algebra" -> "linear algebra"

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

clean up GUI discussion for MATLAB

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
respond to @bjnath comments

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,50 @@ Some Key Differences
+       NumPy's. The syntax for basic matrix operations is nice and clean, but the
+       API for adding GUIs and making full-fledged applications is more or less
+       an afterthought.
+     - NumPy is  based on Python, which was designed from the outset to be an
+       excellent general-purpose programming language.  The advantage to NumPy
+       is access to Python libraries including: `SciPy
+       <https://www.scipy.org/>`_, `Matplotlib <https://matplotlib.org/>`_,
+       `Pandas <https://pandas.pydata.org/>`_, `OpenCV <https://opencv.org/>`_,
+       and more. In addition to Python-specific packages, much `scientific
+       computing software
+       <https://en.wikipedia.org/wiki/List_of_Python_software#Embedded_as_a_scripting_language>`_
+       use Python as a scripting language. 
```
You don't have to word it exactly this way, but I thought since this was a guide to newcomers it would be helpful to give a short description of these packages. 

Note the lower-case spelling of matplotlib and pandas.

The Wikipedia entry seemed off-topic (it lists programs that can call out to Python rather than packages a Python program can call).


```suggestion
     - Because NumPy is based on Python, a general-purpose programming 
     language, users have access to the entire Python "scientific ecosystem," 
     whose packages include the computational tools of `SciPy
       <https://www.scipy.org/>`_, the plotting software of 
       `matplotlib <https://matplotlib.org/>`_,
      the data manipulation and analysis tool `pandas <https://pandas.pydata.org/>`_, 
      the image-processing library`OpenCV <https://opencv.org/>`_, and more.
```

eric-wieser replied:
> The Wikipedia entry seemed off-topic (it lists programs that can call out to Python rather than packages a Python program can call).

Agreed. https://scipy.org/topical-software.html may be a better link.

cooperrc replied:
> > The Wikipedia entry seemed off-topic (it lists programs that can call out to Python rather than packages a Python program can call).
> 
> Agreed. https://scipy.org/topical-software.html may be a better link.

The scipy.org is a good link to see where Python is used to build software. 

> The Wikipedia entry seemed off-topic (it lists programs that can call out to Python rather than packages a Python program can call).

I like the wiki entry because it lists scientific computing software that use Python as a scripting language. Which means you can create NumPy-based scripts to run FEA/CFD/dynamics/etc. without having to build it yourself. Then, you can post-process the data. We can drop it if its not of interest to the broader MATLAB conversions, but its one of the benefits people coming from MATLAB might not consider at first.

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,50 @@ Some Key Differences
+
+   * - In MATLAB, arrays have pass-by-value semantics, with a lazy
+       copy-on-write scheme to prevent creating copies until they
        are actually needed.  Slice operations copy parts of the array.
```
```suggestion
       are needed.  Slice operations copy parts of the array.
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,50 @@ Some Key Differences
+       copy-on-write scheme to prevent creating copies until they
        are actually needed.  Slice operations copy parts of the array.
-     - In NumPy arrays have pass-by-reference semantics.  Slice operations
+     - In NumPy, arrays have pass-by-reference semantics.  Slice operations
        are views into an array.
```
We should qualify this a little  (perhaps "arrays **often** have"). Slices are indeed views, but some operations return copies.

bjnath replied:
@eric-wieser "Pass by reference" isn't mine.

cooperrc replied:
Does this need a rewrite? How about:

`MATLAB array slicing uses pass-by-value semantics, with a lazy copy-on-write scheme to prevent creating copies until they are needed. Slice operation copy parts of the array.`

`NumPy array slicing uses pass-by-reference, that does not copy the arguments. Slice operations are views into an array.`

cooperrc replied:
Updated to the rewrite in a3b1c6e303d6a1261b6068f4d1e406c94d178e7a.

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,50 @@ Some Key Differences
-The ``array`` is thus much more advisable to use.  Indeed, we intend to
-deprecate ``matrix`` eventually.
 
 Table of Rough MATLAB-NumPy Equivalents
```
They can see it's a table and the page is about MATLAB vs NumPy. 

Headings use lower case (so "equivalents," not "Equivalents").
```suggestion
Rough equivalents
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,50 @@ Some Key Differences
 =======================================
 
-The table below gives rough equivalents for some common MATLAB®
+The table below gives rough equivalents for some common MATLAB
 expressions. **These are not exact equivalents**, but rather should be
 taken as hints to get you going in the right direction. For more detail
 read the built-in documentation on the NumPy functions.
```
```suggestion
This table gives rough equivalents for some common MATLAB
expressions. These are workalikes, **not exact equivalents**, so 
read the documentation.
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
-     - ``ndim(a)`` or ``a.ndim``
-     - get the number of dimensions of an array
+     - ``np.ndim(a)`` or ``a.ndim``
+     - get the number of dimensions of the array, ``a``
```
```suggestion
     - number of dimensions of array ``a``
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
-     - ``size(a)`` or ``a.size``
-     - get the number of elements of an array
+     - ``np.size(a)`` or ``a.size``
+     - get the number of elements of the array, ``a``
```
```suggestion
     - number of elements of array ``a``
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
-     - ``shape(a)`` or ``a.shape``
-     - get the "size" of the matrix
+     - ``np.shape(a)`` or ``a.shape``
+     - get the "size" of the array, ``a``
```
```suggestion
     - "size" of array ``a``
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
 
    * - ``size(a,n)``
      - ``a.shape[n-1]``
      - get the number of elements of the n-th dimension of array ``a``. (Note
```
```suggestion
     - number of elements of the n-th dimension of array ``a``. (Note
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
      - ``a.shape[n-1]``
      - get the number of elements of the n-th dimension of array ``a``. (Note
-       that MATLAB® uses 1 based indexing while Python uses 0 based indexing,
+       that MATLAB uses 1 based indexing while Python uses 0 based indexing,
```
```suggestion
       that MATLAB uses 1-based indexing while Python uses 0-based indexing,
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
    * - ``a(2,5)``
      - ``a[1,4]``
-     - access element in second row, fifth column
+     - access element in second row, fifth column in a 2D array, ``a``
```
```suggestion
     - access element in second row, fifth column in 2D array ``a``
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
    * - ``a(2,:)``
      - ``a[1]`` or  ``a[1,:]``
-     - entire second row of ``a``
+     - entire second row of a 2D array, ``a``
```
```suggestion
     - entire second row of 2D array ``a``
```

eric-wieser replied:
This isn't really gramatically reasonable any more, even as a fragment - we can drop the articles at the start of the description, but it feels wrong to do so mid-description.

bjnath replied:
@eric-wieser  A grammatically parallel construction is  "Footprint of Apollo astronaut Neil Armstrong."

bjnath replied:
It's apposition, is legal, and nobody thinks twice.  "Sister of advice columnist Ann Landers." "Cover of book by NumPy creator Travis Oliphant." 

Consider alternatives:

<br>

> Footprint of the Apollo astronaut Neil Armstrong

Only if we're distinguishing it from "Footprint of the Indiana lawyer Neil Armstrong."

<br>

> Footprint of the Apollo astronaut, Neil Armstrong

This says there was only one Apollo astronaut.

<br>

> Footprint of an Apollo astronaut, Neil Armstrong

This suggests we're contrasting "Footprint of a Gemini astronaut, Gus Grissom."

<br>

> Footprint of an Apollo astronaut Neil Armstrong

Bad usage.

<br>

> Footprint of Apollo astronaut, Neil Armstrong

Bad usage.

<br>

"Footprint of Apollo astronaut Neil Armstrong" is in fact the best way to put it in this context, as is "entire second row of 2D array ``a``."

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
    * - ``a(1:5,:)``
      - ``a[0:5]`` or ``a[:5]`` or ``a[0:5,:]``
-     - the first five rows of ``a``
+     - the first five rows of a 2D array, ``a``
```
```suggestion
     - first five rows of 2D array ``a``
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
    * - ``a(end-4:end,:)``
      - ``a[-5:]``
-     - the last five rows of ``a``
+     - the last five rows of a 2D array, ``a``
```
```suggestion
     - last five rows of 2D array ``a``
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -258,63 +187,63 @@ Linear Algebra Equivalents
-     - rows one to three and columns five to nine of ``a``.  This gives
-       read-only access.
+     - ``a[0:3,4:9]``
+     - rows one to three and columns five to nine of a 2D array, ``a``. 
```
We can't say "row one" if the two notations disagree on what row 1 is. We need to stick to ordinals or draw a picture.

eric-wieser replied:
@bjnath, do you prefer this?
```suggestion
     - The first through third rows and fifth through ninth columns of a 2D array, ``a``. 
```

bjnath replied:
@eric-wieser Yes, this is what I had in mind.

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,67 +479,76 @@ Linear Algebra Equivalents
-     - cholesky factorization of a matrix (``chol(a)`` in matlab returns an
+   * - ``c=chol(a)`` where ``a==c'*c``
+     - ``c = linalg.cholesky(a)`` where ``a == c@c.T``
+     - cholesky factorization of a matrix (``chol(a)`` in MATLAB returns an
```
```suggestion
     - Cholesky factorization of a matrix (``chol(a)`` in MATLAB returns an
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,67 +479,76 @@ Linear Algebra Equivalents
-     - sort the rows of the matrix
+   * - ``sort(a,2)``
+     - ``np.sort(a,axis=1)`` or ``a.sort(axis=1)``
+     - sort the each row of 2D matrix, ``a``
```
```suggestion
     - sort each row of 2D matrix, ``a``
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,67 +479,76 @@ Linear Algebra Equivalents
-     - ``unique(a)``
-     -
+     - ``np.unique(a)``
+     - returns a vector of unique values in array ``a``
```
```suggestion
     - return a vector of unique values in array ``a``
```

eric-wieser replied:
Or even just
```suggestion
     - a vector of unique values in array ``a``
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,67 +479,76 @@ Linear Algebra Equivalents
    * - ``squeeze(a)``
      - ``a.squeeze()``
-     -
+     - remove singleton dimensions of array ``a`` Note that MATLAB will always
```
```suggestion
     - remove singleton dimensions of array ``a``. Note that MATLAB will always
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -613,7 +564,7 @@ but the commands ``help`` and ``source`` will usually list the filename
 ``import inspect``) which provides a ``getfile`` that often works.
 
-\ **INDEXING**: MATLAB® uses one based indexing, so the initial element
+\ **INDEXING**: MATLAB uses one based indexing, so the initial element
```
```suggestion
\ **INDEXING**: MATLAB uses 1-based indexing, so the initial element
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -613,7 +564,7 @@ but the commands ``help`` and ``source`` will usually list the filename
 
-\ **INDEXING**: MATLAB® uses one based indexing, so the initial element
+\ **INDEXING**: MATLAB uses one based indexing, so the initial element
 of a sequence has index 1. Python uses zero based indexing, so the
```
```suggestion
of a sequence has index 1. Python uses 0-based indexing, so the
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -623,55 +574,176 @@ indexing <https://groups.google.com/group/comp.lang.python/msg/1bf4d925dfbf368?q
 See also `a text by prof.dr. Edsger W.
 Dijkstra <https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html>`__.
```
Please delete the entire section beginning "Confusion and flamewars"
```
initial element of a sequence has index 0. Confusion and flamewars arise
```
and ending with 
```
Dijkstra <https://www.cs.utexas.edu/users/EWD/transcriptions/EWD08xx/EWD831.html>`__.
```

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -696,26 +768,39 @@ this is just an example, not a statement of "best practices"):
+<https://en.wikipedia.org/wiki/List_of_Python_software#Embedded_as_a_scripting_language>`_
+for a list of Scientific packages that use Python as a scripting language
+
+MATLAB and SimuLink® are registered trademarks of The MathWorks.
```
```suggestion
MATLAB and SimuLink® are registered trademarks of MathWorks.
```

eric-wieser replied:
Let's keep the (r) on matlab too

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,67 +479,76 @@ Linear Algebra Equivalents
+     - sort each column of a 2D matrix, ``a``
+
+   * - ``sort(a,2)``
+     - ``np.sort(a,axis=1)`` or ``a.sort(axis=1)``
```
PEP8
```suggestion
     - ``np.sort(a, axis=1)`` or ``a.sort(axis=1)``
```

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,67 +479,76 @@ Linear Algebra Equivalents
-     - ``I = argsort(a[:,i]), b=a[I,:]``
-     - sort the rows of the matrix
+   * - ``[b,I]=sortrows(a,1)``
+     - ``I=np.argsort(a[:,0]); b=a[I,:]``
```
PEP8
```suggestion
     - ``I = np.argsort(a[:,0]); b = a[I,:]``
```

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,67 +479,76 @@ Linear Algebra Equivalents
-   * - ``regress(y,X)``
-     - ``linalg.lstsq(X,y)``
-     - multilinear regression
+   * - ``Z\y``
+     - ``linalg.lstsq(Z,y)``
+     - perform a linear regression of the form :math:`\mathbf{Zx}=\mathbf{y}`
```
```suggestion
   * - ``x = Z\y``
     - ``x = linalg.lstsq(Z,y)``
     - perform a linear regression of the form :math:`\mathbf{Zx}=\mathbf{y}`
```
otherwise it's not clear what `x` is

eric-wieser replied:
Or maybe
```suggestion
   * - ``Z\y``
     - ``linalg.lstsq(Z,y)``
     - the solution :math:`x` to a linear regression of the form :math:`\mathbf{Zx}=\mathbf{y}`
```

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

remove redundant wording L62

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update table heading to Rough equivalents

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -664,10 +615,131 @@ possible to do this fairly efficiently. Note that the scan order used by
+The ``array`` is thus much more advisable to use.  Indeed, we intend to
+deprecate ``matrix`` eventually.
+
 Customizing Your Environment
```
```suggestion
Customizing your environment
```

bjnath commented:
Please fix the headline case (also note the hyphen in the first one):

General Purpose Equivalents -> General-purpose equivalents
Linear Algebra Equivalents -> Linear algebra equivalents
Customizing Your Environment -> Customizing your environment

bjnath commented:
Please have another look at this sentence:

> NumPy is based on Python, which was designed from the outset to be an excellent general-purpose programming language. 

* "which was designed from the outset":  Don't need "which was." Designed "at the outset" is redundant.
* "designed to be an excellent":  Not useful. Most languages are designed to be excellent. It's only noteworthy if it was designed to be lousy.

bjnath commented:
Please have another look at 
> These are not exact equivalents, but rather should be taken as hints to get you going in the right direction. 

* "but rather should be taken as hints": Just saying they're hints does the job.
* "to get you going in the right direction": Again, readers can safely assume we're giving good hints, and we only need to tell them when we give bad ones.

How about
> These are hints, not exact equivalents

or my original more precise suggestion

> These are workalikes, not exact equivalents.

bjnath commented:
> For more detail read the built-in documentation on the NumPy functions.

These users are new to NumPy and might be grateful for [a link](https://numpy.org/doc/stable/reference/routines.html).

Also, what exactly is "built-in" documentation?

bjnath commented:
On reflection, rather than trying to accomplish everything now, perhaps it's better to review only technical issues (including code styling) in this PR and address remaining verbal ones in a follow-up PR.

bjnath commented:
> MATLAB® and NumPy have a lot in common, but NumPy was created to work with
> Python, not to be a MATLAB clone.  This guide will help MATLAB users get started
> with NumPy. **NumPy and SciPy were created to do numerical and scientific
> computing in the most natural way with Python, not to be MATLAB clones.  This
> page is intended to be a place to collect wisdom about the differences, mostly
> for the purpose of helping proficient MATLAB users become proficient NumPy and
> SciPy users.**
> 

In the opening paragraph, the old text  (last two sentences) needs to be deleted.

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

add description of element-wise multiplication in intro

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
incorporate @bjnath comments

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

PEP8

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

PEP8

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

define `x` in linear regression

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

fix title on L746

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
add even for scalars to MATLAB ND array comparison

cooperrc commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -25,165 +25,51 @@ Some Key Differences
+       use the ``@`` operator in Python 3.5 and above).
+
+   * - MATLAB uses 1 (one) based indexing. The initial element of a
        sequence is found using a(1).
```
I like this change, but I made one small edit
`MATLAB numbers indexes from 1; a(1) is the first element.`
to
`MATLAB numbers indices from 1; a(1) is the first element.`

bjnath replied:
Thanks for making the change.

The [Google style guide](https://developers.google.com/style/word-list#letter-i) has a rule for the plural of **index**:

> **index**
>     Use the plural indexes unless there is a domain-specific reason (for example, a mathematical or financial context) to use 

I think this is an "indexes" situation; calling it a mathematical context is a stretch. 

If our docs uniformly used "indices" we'd toss the rule, but that's not the case.

The two or three other "indices" instances on the page, which I overlooked, ought to be changed as well.

eric-wieser replied:
Given numpy has a function `np.indices`, I think we should use that spelling everywhere.

cooperrc replied:
> Given numpy has a function `np.indices`, I think we should use that spelling everywhere.

@bjnath, Thanks for the Google style guide. I think keeping the indices and `np.indices` would be nice to keep consistent with the numpy function. I would classify this as mathematical context, because we are discussing arrays and matrices. 

After this discussion, I noticed one other "indexes" -> "indices" on L580 in the __Notes__. I'll push the change with next commit.

bjnath replied:
I loathe "indices" but others equally loathe "indexes," and I have no good counter for `np.indices`, which I guess goes back as far as one can go.  

I'll find a place to collect style rules that supplement or override Google's.

cooperrc commented:
> Please have another look at
> 
> > These are not exact equivalents, but rather should be taken as hints to get you going in the right direction.
> 
> * "but rather should be taken as hints": Just saying they're hints does the job.
> * "to get you going in the right direction": Again, readers can safely assume we're giving good hints, and we only need to tell them when we give bad ones.
> * "exact equivalents": Equivalents are already exact and don't need the modifier.
> 
> That leaves
> 
> > These are **hints**, not equivalents.
> 
> or how about
> 
> > These are workalikes, not equivalents.

How about:

`These are similar expressions, not equivalents.`

cooperrc committed:
fix heading case

cooperrc committed:
remove extra sentences in Intro

bjnath commented:
> How about:
>> These are similar expressions, not equivalents.

Sounds great!

cooperrc commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -277,36 +206,36 @@ Linear Algebra Equivalents
    * - ``[ 1 2 3; 4 5 6 ]``
      - ``array([[1.,2.,3.], [4.,5.,6.]])``
-     - 2x3 matrix literal
+     - 2D array defined as 2x3 matrix
```
After you pointed it out, I don't like it either. I'm replacing `matrix` with `2D array`, since there are places that sometimes refer to matrix as `2D matrix` and other times just `matrix`. 

Both MATLAB and NumPy deal with arrays, so I think its best to keep the `array` verbage.

cooperrc committed:
PEP8 corrections and @melissawm edits

cooperrc commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -231,13 +134,39 @@ General Purpose Equivalents
+         elif a == 1:
+             print('a=1')
+
+     - create an if-else statement to check if ``a`` is 0 or 1 and print result
```
That's a great point that could be very confusing. I changed the values to `a=4`, `a == 4` and `a == 5`. 

I added a row above to reference the Python boolean:

     - ::
        >> 4 == 4
        ans = 1
        >> 4 == 5
        ans = 0

     - ::

        >>> 4 == 4
        True
        >>> 4 == 5
        False


The [boolean objects](https://docs.python.org/3/library/stdtypes.html?highlight=boolean#boolean-values) in Python are ``True`` and ``False``, as opposed to MATLAB logical types of ``1`` and ``0``.

cooperrc commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -206,21 +92,38 @@ General Purpose Equivalents
+         for i in range(1, 4):
+            print(i)
+
+     - use a for-loop to print the numbers 1, 2, and 3
```
I added the link to [range](https://docs.python.org/3/library/functions.html#func-range)

> use a for-loop to print the numbers 1, 2, and 3 using [range](https://docs.python.org/3/library/functions.html#func-range)

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
-read the built-in documentation on the NumPy functions.
+expressions. These are similar expressions, not equivalents. For
+details, see the `documentation
+<https://numpy.org/doc/stable/reference/routines.html>`_. 
```
This should be an internal Sphinx. I don't remember the exact syntax...

cooperrc replied:
So the Routines index.html start with reference:
`.. _reference:`
But using this `documentation <reference>` can't find that index page. I'm having a bit of trouble figuring out the rst cross-refs. 

@bjnath and @melissawm do you know how to get that cross-ref?

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -231,23 +136,68 @@ General Purpose Equivalents
+        False
+
+     - The `boolean objects
+       <https://docs.python.org/3/library/stdtypes.html?highlight=boolean#boolean-values>`_
```
Should be able to use an intersphinx link here, perhaps `<python:boolean-values>`

cooperrc replied:
> Should be able to use an intersphinx link here, perhaps `<python:boolean-values>`

Nice! That worked, I'll push the change on next commit.

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -231,23 +136,68 @@ General Purpose Equivalents
+     - ``io.loadmat('data.mat')``
+     - Load MATLAB variables saved to the file ``data.mat``. (Note: if saving files from
+       MATLAB/Octave use a recent binary format, Python's ``loadmat`` will
+       create a dictionary with arrays and further information)
```
(on mobile so can't select multiple lines)

There's something wrong with the grammar here. Are you trying to say "if saving from matlab using...", or "if saving from matlab, use ..."?

cooperrc replied:
trying to specify that if you're going to save arrays into the `data.mat` you have to use a binary format in MATLAB/Octave. 

how about:
`(Note: when saving arrays to ``data.mat`` in MATLAB/Octave, use a recent binary format. Python's ``loadmat`` will create a dictionary with the saved arrays and further information.)`

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
+         for i in range(1, 4):
+            print(i)
+
+     - use a for-loop to print the numbers 1, 2, and 3 using `range
```
```suggestion
     - use a for-loop to print the numbers 1, 2, and 3 using :func:`range`
```
And delete the url below.

cooperrc replied:
> And delete the url below.

my html did not generate a link with this change. Is there a way to specify Python instead of NumPy function?

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -696,26 +789,39 @@ this is just an example, not a statement of "best practices"):
 Links
 =====
 
 See http://mathesaurus.sf.net/ for another MATLAB/NumPy
```
This site is rather out of date regarding both python and numpy versions. Perhaps we should indicate that it is outdated.

cooperrc replied:
That's true, we can specify that its out of date. 

As an alternative link, there's a RealPython article on [MATLAB vs Python](https://realpython.com/matlab-vs-python/), maybe we could add this in the references.

eric-wieser replied:
I'd be fine for now with just 
```
Another somewhat outdated MATLAB/NumPy cross-reference can be found at http://mathesaurus.sf.net/
```

cooperrc replied:
`Another somewhat outdated MATLAB/NumPy cross-reference can be found at http://mathesaurus.sf.net/`

updated reference in commit: de298885a4725889512e8172534d134b630c4da6

bjnath commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
+       linear algebra. 
+
+     - In NumPy, the basic type is a multidimensional ``array``.  Array
+       assignments in NumPy are usually stored as ND arrays with the
```
```suggestion
       assignments in NumPy are usually stored as n-dimensional arrays with the
```
`ND`, while consistent, might be cryptic. I think it's reasonable to say `n-dimensional`. I apologize if I steered you in the wrong direction.

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

revert ND to n-dimensional, much clearer

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
updating refs and some formatting

cooperrc committed:
Merge branch 'update_numpy4matlab' of https://github.com/cooperrc/numpy into update_numpy4matlab

mattip commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,124 @@ NumPy for MATLAB users
+Some key differences
 ====================
 
 .. list-table::
 
```
I think this will enable the formatting in the `.. raw` directive on line 14 (if I got the blank line and the number of spaces before the `:class:`)

```suggestion
.. list-table::
   :class: docutils

```

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

Co-authored-by: Matti Picus <matti.picus@gmail.com>

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
+       is access to Python libraries including: `SciPy
+       <https://www.scipy.org/>`_, `Matplotlib <https://matplotlib.org/>`_,
+       `Pandas <https://pandas.pydata.org/>`_, `OpenCV <https://opencv.org/>`_,
+       and more. In addition to Python-specific packages, much `scientific
+       computing software
+       <https://en.wikipedia.org/wiki/List_of_Python_software#Embedded_as_a_scripting_language>`_
+       use Python as a scripting language. 
```
Looking at that list, not a particularly large fraction of the applications strike me as "scientific computing". I'd be inclined to either throw out this sentence, or phrase it as:
```suggestion
       and more. In addition, Python is often `embedded as a scripting language
       <https://en.wikipedia.org/wiki/List_of_Python_software#Embedded_as_a_scripting_language>`_
       in other software, allowing NumPy to be used there too. 
```

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
 
    * - ``which func``
-     - `see note HELP <numpy-for-matlab-users.notes>`__
+     - `see note HELP <numpy-for-matlab-users.notes>`_
```
This doesn't appear to work properly. I can only assume the `__` was necessary.

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
 
    * - ``type func``
-     - ``source(func)`` or ``func??`` (in Ipython)
+     - ``source(func)`` or ``func??`` (in IPython)
```
```suggestion
     - ``inspect.source(func)`` or ``func??`` (in IPython)
```

eric-wieser replied:
I think this was meant to say
```suggestion
     - ``np.source(func)`` or ``func??`` (in IPython)
```
where `np.source` is essentially just a thin wrapper around `inspect.getsource`

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -231,23 +136,68 @@ General Purpose Equivalents
+   * - ``load data.mat``
+     - ``io.loadmat('data.mat')``
+     - Load MATLAB variables saved to the file ``data.mat``. (Note: When saving arrays to
+       ``data.mat`` in MATLAB/Octave, use a recent binary format. Python's ``loadmat``
```
```suggestion
       ``data.mat`` in MATLAB/Octave, use a recent binary format. :func:`scipy.io.loadmat`
```

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
+         for i in range(1, 4):
+            print(i)
+
+     - use a for-loop to print the numbers 1, 2, and 3 using :func:`range`
```
```suggestion
     - use a for-loop to print the numbers 1, 2, and 3 using :py:class:`range <range>`
```

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
+
    * - ``a && b``
      - ``a and b``
      - short-circuiting logical  AND operator (Python native operator);
```
```suggestion
   - short-circuiting logical AND operator (:ref:`Python native operator <python:boolean>`);
```

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -231,23 +136,68 @@ General Purpose Equivalents
      - short-circuiting logical OR operator (Python native operator);
```
- short-circuiting logical OR operator (:ref:`Python native operator <python:boolean>`);

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -231,23 +136,68 @@ General Purpose Equivalents
+        False
+
+     - The `boolean objects
+       <python:boolean-values>`_
```
```suggestion
       The :ref:`boolean objects <python:bltin-boolean-values>`
```

melissawm commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -537,76 +500,85 @@ Linear Algebra Equivalents
+do: ``ind=[1, 3]; a[np.ix_(ind, ind)] += 100``.
 
 \ **HELP**: There is no direct equivalent of MATLAB's ``which`` command,
 but the commands ``help`` and ``source`` will usually list the filename
```
I'm pretty sure `source` does not work anymore. The best option is `inspect.getfile(func)`, I think.

eric-wieser replied:
```suggestion
but the commands :func:`help`` and :func:`numpy.source` will usually list the filename
```

melissawm reviewed (changes requested):
I think I got all the links. This is shaping up really nicely, thanks so much for your patience @cooperrc ! I think we're almost there :)

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update the loadmat reference

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update the documentation reference

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

qualify use of NumPy in scripting languages

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update ref for see note HELP L104

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update range reference

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update Python boolean ref

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

give link for boolean in the short-circuit logical AND

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

cooperrc commented:
@melissawm [`inspect.getsource` vs `source` - comment](https://github.com/numpy/numpy/pull/17159#discussion_r486678779) 

I run into a `TypeError` if the source code can't be retrieved using `inspect.getsource`, but when you run `source` you just get a message, 

`Not available for this object`

I'm comparing functions, `array` and `ones` in my case.

cooperrc commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -7,220 +7,125 @@ NumPy for MATLAB users
+   :class: docutils
+
+   * - In MATLAB, the basic type, even for scalars, is a
+       multidimensional array. Array assignments in MATLAB are stored as
```
I think this makes more sense on L34, when we reference n-dimensional arrays in NumPy. 
```
assignments in NumPy are usually stored as :ref:`n-dimensional arrays <arrays>` 
```

cooperrc replied:
Added the reference to `<arrays>` on L34 in commit: de298885a4725889512e8172534d134b630c4da6

eric-wieser commented:
@cooperrc, are you referring to `np.source`? Otherwise, I don't know what `source` function you're using.

cooperrc commented:
> @cooperrc, are you referring to `np.source`? Otherwise, I don't know what `source` function you're using.

Yes, sorry about that: `np.source`

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -354,30 +303,30 @@ Linear Algebra Equivalents
      - matrix whose i,jth element is (a_ij > 0.5).  The MATLAB result is an
-       array of 0s and 1s.  The NumPy result is an array of the boolean
-       values ``False`` and ``True``.
+       array of logical types 0 and 1.  The NumPy result is an array of the boolean
+       types ``False`` and ``True``.
```
```suggestion
       array of logical values 0 and 1.  The NumPy result is an array of the boolean
       values ``False`` and ``True``.
```
Either "of boolean type" or "of boolean values False and True"

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -386,112 +335,125 @@ Linear Algebra Equivalents
 
    * - ``[1:10]'``
-     - ``arange(1.,11.)[:, newaxis]``
+     - ``np.arange(1.,11.)[:, newaxis]``
```
```suggestion
     - ``np.arange(1.,11.)[:, np.newaxis]``
```

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update `source` -> `np.source` L108

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update link for boolean

Co-authored-by: Melissa Weber Mendonça <melissawm@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update links for `help` and `np.source` L583

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

update references to boolean values instead of types

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

add link to cholesky

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

add missin `np.` to L360

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

Co-authored-by: Ben Nathanson <github@bigriver.xyz>

cooperrc committed:
Apply suggestions from code review

incorporate @bjnath edits, nice suggestions

Co-authored-by: Ben Nathanson <github@bigriver.xyz>
Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

cooperrc committed:
checked links and formatting

mattip committed:
BUILD: pin pygments to 2.6.1

cooperrc committed:
updated pass-by-reference discussion in array slicing

cooperrc committed:
updated pass-by-reference discussion in array slicing

cooperrc committed:
change verb for slicing in pass-by-value MATLAB

cooperrc commented:
@eric-wieser, @bjnath, and @melissawm I think we hit all of the major changes that needed to go into this update. I'm really happy with this revision, thanks so much for your help!

Can we table any further comments/suggestions and merge the PR? If there are other issues, I can open another PR to continue our discussion (specifically our thoughts on some the finer points on array dimensions in NumPy vs MATLAB).

eric-wieser commented on doc/source/user/numpy-for-matlab-users.rst:
```diff
@@ -696,26 +786,39 @@ this is just an example, not a statement of "best practices"):
+<https://en.wikipedia.org/wiki/List_of_Python_software#Embedded_as_a_scripting_language>`_
+for a list of softwares that use Python as a scripting language
+
+MATLAB®  and SimuLink® are registered trademarks of MathWorks.
```
Let's put this back as it was, the `The` is what the matlab webpages say
```suggestion
MATLAB® and SimuLink® are registered trademarks of The MathWorks, Inc.
```

cooperrc committed:
Update doc/source/user/numpy-for-matlab-users.rst

revert reference for MathWorks to original -> "The MathWorks, Inc." as seen on website

Co-authored-by: Eric Wieser <wieser.eric@gmail.com>

mattip commented:
Thanks @cooperrc and all the reviewers.
